src/text.c

```diff
@@ -0,0 +1,358 @@
+/* sdb - MIT - Copyright 2020 - thestr4ng3r */
+
+#include "sdb.h"
+
+/**
+ * ********************
+ * Plaintext SDB Format
+ * ********************
+ *
+ * Files are UTF-8 and use '\n' line endings. Always.
+ * 
+ * Lines starting with '/' denote the path of the namespace for the following data:
+ *
+ *   /some/namespace
+ * 
+ * The default path is root, just a slash also means root.
+ * These paths are always absolute from the root. Characters that must be escaped in a path are: '/', '\\', '\n':
+ *
+ *   /s\/ome/name\nspa\\ce
+ *
+ * SDB entries are written each as a single k=v line:
+ *
+ *   somekey=somevalue
+ *
+ * To distinguish these from path lines, if there is a leading '/' in the key, it must be escaped
+ * (slashes later in the line don't have to be escaped):
+ *
+ *   \/slashedkey=somevalue
+ *
+ * Other than that, at any postion, '\\' and '\n' must be escaped:
+ *   
+ *   some\\key=some\nvalue
+ *
+ * In the key, '=' must also be escaped (not necessary in the value):
+ *
+ *   some\=key=some=value
+ * 
+ * --------
+ * Example:
+ *
+ *   /
+ *   key=intheroot
+ *   \/slashedkey=somevalue
+ *   some\\key=some\nvalue
+ *   some\=key=some=value
+ *
+ *   /subns
+ *   some=stuff in the sub-namespace
+ *
+ *   /subns/deeper
+ *   this=is in /subns/deeper
+ *
+*/
+
+static int cmp_ns(const void *a, const void *b) {
+	const SdbNs *nsa = a;
+	const SdbNs *cia = b;
+	return strcmp (nsa->name, cia->name);
+}
+
+static bool text_fsave(Sdb *s, FILE *f, bool sort, SdbList *path) {
+// n = position we are currently looking at
+// p = position until we have already written everything
+// macro to flush a block of text that doesn't have to be escaped
+#define FLUSH do { if (p != n) { fwrite (p, 1, n - p, f); p = n; } } while (0)
+// macro to flush and skip a char
+#define SKIP do { FLUSH; p++; } while (0)
+	// path
+	fwrite ("/", 1, 1, f); // always print a /, even if path is empty
+	SdbListIter *it;
+	const char *path_token;
+	bool first = true;
```

Comment:
this `first` thing can probably be avoided?

Reply:
No, I always need to print a slash, even when path is empty. So I have to do it outside the loop and skip it in the first loop iteration, which is what the `first` is needed for.